File: apps/webhook_server/src/sharpedge_webhooks/routes/v1/portfolio.py
"""GET /api/v1/users/{user_id}/portfolio — auth-gated portfolio stats."""

from __future__ import annotations

from collections import defaultdict
from decimal import Decimal

from fastapi import APIRouter, HTTPException, status

from sharpedge_db.queries.users import (
    get_internal_user_id_by_supabase_auth,
    get_unit_size_for_user,
)
from sharpedge_webhooks.routes.v1.deps import CurrentUser  # noqa: TC001

router = APIRouter(tags=["v1"])


def _is_pending_result(value: object) -> bool:
    return str(value or "").upper() == "PENDING"


def get_performance_summary(user_id: str):
    """Import-time lazy shim — real import happens at call time."""
    from sharpedge_db.queries.bets import get_performance_summary as _fn

    return _fn(user_id=user_id)


def get_user_bets_history(user_id: str, limit: int = 200):
    """Import-time lazy shim — real import happens at call time."""
    from sharpedge_db.queries.bets import get_user_bets_history as _fn

    return _fn(user_id=user_id, limit=limit)


def load_portfolio_breakdowns(internal_user_id: str) -> dict:
    """Aggregate settled-bet slices for the portfolio API (test patch target)."""
    from sharpedge_db.queries.bets import (
        get_breakdown_by_bet_type,
        get_breakdown_by_juice_bucket,
        get_breakdown_by_sport,
        get_breakdown_by_sportsbook,
    )

    sports = get_breakdown_by_sport(internal_user_id)
    types = get_breakdown_by_bet_type(internal_user_id)
    books = get_breakdown_by_sportsbook(internal_user_id)
    juice = get_breakdown_by_juice_bucket(internal_user_id)
    return {
        "by_sport": [
            {
                "sport": str(s.sport),
                "total_bets": s.total_bets,
                "wins": s.wins,
                "losses": s.losses,
                "win_rate": float(s.win_rate),
                "roi": float(s.roi),
            }
            for s in sports
        ],
        "by_bet_type": [
            {
                "bet_type": str(t.bet_type),
                "total_bets": t.total_bets,
                "wins": t.wins,
                "losses": t.losses,
                "win_rate": float(t.win_rate),
                "roi": float(t.roi),
            }
            for t in types
        ],
        "by_book": books,
        "by_juice": juice,
    }


@router.get("/users/{user_id}/portfolio")
async def portfolio(user_id: str, current_user: CurrentUser) -> dict:
    """Return portfolio stats for a user. Requires valid Supabase JWT.

    RLS ensures Supabase only returns data belonging to the authenticated user.
    We additionally verify the requested user_id matches the JWT user_id.
    """
    if current_user["id"] != user_id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Access denied: cannot view another user's portfolio",
        )

    internal_user_id = get_internal_user_id_by_supabase_auth(user_id)
    if not internal_user_id:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="User profile not found — complete sign-up or contact support",
        )

    summary = get_performance_summary(user_id=internal_user_id)
    bets = get_user_bets_history(user_id=internal_user_id, limit=500)
    active_bets = [b for b in bets if _is_pending_result(b.get("result"))]
    settled_bets = [b for b in bets if not _is_pending_result(b.get("result"))]

    # CLV average from settled bets with clv field
    clv_values = [float(b["clv"]) for b in bets if b.get("clv") is not None]
    clv_average = sum(clv_values) / len(clv_values) if clv_values else 0.0

    # Running max drawdown
    running_balance = 0.0
    peak = 0.0
    max_drawdown = 0.0
    for bet in reversed(settled_bets):
        running_balance += float(bet.get("profit") or 0)
        if running_balance > peak:
            peak = running_balance
        drawdown = peak - running_balance
        if drawdown > max_drawdown:
            max_drawdown = drawdown

    # Time-series history: group settled bets by month (oldest first)
    # Bucket key: "YYYY-MM"
    by_month: dict[str, dict] = defaultdict(lambda: {"profit": 0.0, "stake": 0.0})
    for bet in reversed(settled_bets):  # oldest first
        ts = bet.get("settled_at") or bet.get("placed_at") or ""
        if not ts:
            continue
        month_key = ts[:7]  # "YYYY-MM"
        by_month[month_key]["profit"] += float(bet.get("profit") or 0)
        by_month[month_key]["stake"] += float(bet.get("stake") or 0)

    roi_history = []
    bankroll_history = []
    cum_profit = 0.0
    cum_stake = 0.0
    for month_key in sorted(by_month):
        cum_profit += by_month[month_key]["profit"]
        cum_stake += by_month[month_key]["stake"]
        cum_roi = (cum_profit / cum_stake * 100) if cum_stake else 0.0
        # Format label: "Jan '25"
        try:
            from datetime import datetime

            dt = datetime.strptime(month_key, "%Y-%m")
            label = dt.strftime("%b '%y")
        except ValueError:
            label = month_key
        roi_history.append({"date": label, "roi": round(cum_roi, 2)})
        bankroll_history.append({"date": label, "bankroll": round(cum_profit, 2)})

    unit_size = get_unit_size_for_user(internal_user_id)
    breakdowns = load_portfolio_breakdowns(internal_user_id)

    return {
        "user_id": user_id,
        "roi": float(summary.roi),
        "win_rate": float(summary.win_rate) / 100,
        "clv_average": clv_average,
        "drawdown": max_drawdown,
        "unit_size": float(unit_size.quantize(Decimal("0.01"))) if unit_size > 0 else 0.0,
        **breakdowns,
        "active_bets": [
            {
                "id": b.get("id", ""),
                "event": b.get("game", ""),
                "stake": float(b.get("stake") or 0),
                "book": b.get("sportsbook", ""),
            }
            for b in active_bets
        ],
        "total_bets": summary.total_bets,
        "wins": summary.wins,
        "losses": summary.losses,
        "roi_history": roi_history,
        "bankroll_history": bankroll_history,
    }

File: apps/mobile/lib/models/portfolio.dart
// Parsed GET /api/v1/users/{id}/portfolio (breakdown slices + unit size).

class PortfolioSplitRow {
  final String label;
  final int totalBets;
  final int wins;
  final int losses;
  final double winRate;
  final double roi;

  const PortfolioSplitRow({
    required this.label,
    required this.totalBets,
    required this.wins,
    required this.losses,
    required this.winRate,
    required this.roi,
  });

  static double _num(dynamic v, [double d = 0]) => v is num ? v.toDouble() : d;

  static int _int(dynamic v, [int d = 0]) {
    if (v is int) return v;
    if (v is num) return v.toInt();
    return d;
  }

  static PortfolioSplitRow fromSport(Map<String, dynamic> j) => PortfolioSplitRow(
        label: '${j['sport'] ?? ''}',
        totalBets: _int(j['total_bets']),
        wins: _int(j['wins']),
        losses: _int(j['losses']),
        winRate: _num(j['win_rate']),
        roi: _num(j['roi']),
      );

  static PortfolioSplitRow fromBetType(Map<String, dynamic> j) => PortfolioSplitRow(
        label: '${j['bet_type'] ?? ''}',
        totalBets: _int(j['total_bets']),
        wins: _int(j['wins']),
        losses: _int(j['losses']),
        winRate: _num(j['win_rate']),
        roi: _num(j['roi']),
      );

  static PortfolioSplitRow fromBook(Map<String, dynamic> j) => PortfolioSplitRow(
        label: '${j['sportsbook'] ?? ''}',
        totalBets: _int(j['total_bets']),
        wins: _int(j['wins']),
        losses: _int(j['losses']),
        winRate: _num(j['win_rate']),
        roi: _num(j['roi']),
      );

  static PortfolioSplitRow fromJuice(Map<String, dynamic> j) => PortfolioSplitRow(
        label: '${j['bucket'] ?? ''}',
        totalBets: _int(j['total_bets']),
        wins: _int(j['wins']),
        losses: _int(j['losses']),
        winRate: _num(j['win_rate']),
        roi: _num(j['roi']),
      );
}

class ActivePortfolioBet {
  final String id;
  final String event;
  final double stake;
  final String book;

  const ActivePortfolioBet({
    required this.id,
    required this.event,
    required this.stake,
    required this.book,
  });

  factory ActivePortfolioBet.fromJson(Map<String, dynamic> j) => ActivePortfolioBet(
        id: '${j['id'] ?? ''}',
        event: '${j['event'] ?? ''}',
        stake: (j['stake'] as num?)?.toDouble() ?? 0,
        book: '${j['book'] ?? ''}',
      );
}

class RoiHistoryPoint {
  final String date;
  final double roi;

  const RoiHistoryPoint({required this.date, required this.roi});

  factory RoiHistoryPoint.fromJson(Map<String, dynamic> j) => RoiHistoryPoint(
        date: '${j['date'] ?? ''}',
        roi: (j['roi'] as num?)?.toDouble() ?? 0,
      );
}

class BankrollHistoryPoint {
  final String date;
  final double bankroll;

  const BankrollHistoryPoint({required this.date, required this.bankroll});

  factory BankrollHistoryPoint.fromJson(Map<String, dynamic> j) => BankrollHistoryPoint(
        date: '${j['date'] ?? ''}',
        bankroll: (j['bankroll'] as num?)?.toDouble() ?? 0,
      );
}

class PortfolioSnapshot {
  /// Server sends win rate as 0–1 fraction.
  final double unitSize;
  final double roiPct;
  final double winRateFraction;
  /// Average CLV from settled bets (same units as server).
  final double clvAverage;
  /// Peak-to-trough drawdown in profit dollars (API `drawdown`).
  final double drawdown;
  final int totalBets;
  final int wins;
  final int losses;
  final List<ActivePortfolioBet> activeBets;
  final List<RoiHistoryPoint> roiHistory;
  final List<BankrollHistoryPoint> bankrollHistory;
  final List<PortfolioSplitRow> bySport;
  final List<PortfolioSplitRow> byBetType;
  final List<PortfolioSplitRow> byBook;
  final List<PortfolioSplitRow> byJuice;

  const PortfolioSnapshot({
    required this.unitSize,
    required this.roiPct,
    required this.winRateFraction,
    this.clvAverage = 0,
    this.drawdown = 0,
    this.totalBets = 0,
    this.wins = 0,
    this.losses = 0,
    this.activeBets = const [],
    this.roiHistory = const [],
    this.bankrollHistory = const [],
    required this.bySport,
    required this.byBetType,
    required this.byBook,
    required this.byJuice,
  });

  factory PortfolioSnapshot.fromJson(Map<String, dynamic> j) {
    List<PortfolioSplitRow> mapList(
      dynamic list,
      PortfolioSplitRow Function(Map<String, dynamic>) f,
    ) {
      if (list is! List<dynamic>) return [];
      final out = <PortfolioSplitRow>[];
      for (final e in list) {
        if (e is Map<String, dynamic>) {
          out.add(f(e));
        } else if (e is Map) {
          out.add(f(Map<String, dynamic>.from(e)));
        }
      }
      return out;
    }

    List<ActivePortfolioBet> mapActive(dynamic list) {
      if (list is! List<dynamic>) return [];
      final out = <ActivePortfolioBet>[];
      for (final e in list) {
        if (e is Map<String, dynamic>) {
          out.add(ActivePortfolioBet.fromJson(e));
        } else if (e is Map) {
          out.add(ActivePortfolioBet.fromJson(Map<String, dynamic>.from(e)));
        }
      }
      return out;
    }

    List<RoiHistoryPoint> mapRoi(dynamic list) {
      if (list is! List<dynamic>) return [];
      final out = <RoiHistoryPoint>[];
      for (final e in list) {
        if (e is Map<String, dynamic>) {
          out.add(RoiHistoryPoint.fromJson(e));
        } else if (e is Map) {
          out.add(RoiHistoryPoint.fromJson(Map<String, dynamic>.from(e)));
        }
      }
      return out;
    }

    List<BankrollHistoryPoint> mapBankrollHist(dynamic list) {
      if (list is! List<dynamic>) return [];
      final out = <BankrollHistoryPoint>[];
      for (final e in list) {
        if (e is Map<String, dynamic>) {
          out.add(BankrollHistoryPoint.fromJson(e));
        } else if (e is Map) {
          out.add(BankrollHistoryPoint.fromJson(Map<String, dynamic>.from(e)));
        }
      }
      return out;
    }

    return PortfolioSnapshot(
      unitSize: (j['unit_size'] as num?)?.toDouble() ?? 0,
      roiPct: (j['roi'] as num?)?.toDouble() ?? 0,
      winRateFraction: (j['win_rate'] as num?)?.toDouble() ?? 0,
      clvAverage: (j['clv_average'] as num?)?.toDouble() ?? 0,
      drawdown: (j['drawdown'] as num?)?.toDouble() ?? 0,
      totalBets: PortfolioSplitRow._int(j['total_bets']),
      wins: PortfolioSplitRow._int(j['wins']),
      losses: PortfolioSplitRow._int(j['losses']),
      activeBets: mapActive(j['active_bets']),
      roiHistory: mapRoi(j['roi_history']),
      bankrollHistory: mapBankrollHist(j['bankroll_history']),
      bySport: mapList(j['by_sport'], PortfolioSplitRow.fromSport),
      byBetType: mapList(j['by_bet_type'], PortfolioSplitRow.fromBetType),
      byBook: mapList(j['by_book'], PortfolioSplitRow.fromBook),
      byJuice: mapList(j['by_juice'], PortfolioSplitRow.fromJuice),
    );
  }
}
